### ISSUE TITLE
unpin invalid method parameters

### ISSUE THREAD

**Issue by ryanleecode** [polkadot-api/polkadot-api#489]
Probably related to #488, repro instructions there

---

```log
➜  light-client-experimental git:(json-rpc-effect-bindings) pnpm dlx tsx ./scripts/testing.ts
22:42:24 (Fiber #11) INFO  smoldot Smoldot v2.0.25
22:42:25 (Fiber #12) INFO  smoldot Chain initialization complete for polkadot. Name: "Polkadot". Genesis hash: 0x91b1…90c3. Chain specification starting at: 0x9548…ebd3 (#20408370)
22:42:25 (Fiber #14) INFO  json-rpc-provider -> polkadot JSON RPC provider initialized.
22:42:25 (Fiber #20) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","id":"1-1","method":"rpc_methods","params":[]}
22:42:25 (Fiber #13) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","id":"1-1","result":{"methods":["account_nextIndex","author_hasKey","author_hasSessionKeys","author_insertKey","author_pendingExtrinsics","author_removeExtrinsic","author_rotateKeys","author_submitAndWatchExtrinsic","author_submitExtrinsic","author_unwatchExtrinsic","babe_epochAuthorship","chain_getBlock","chain_getBlockHash","chain_getFinalizedHead","chain_getHeader","chain_subscribeAllHeads","chain_subscribeFinalizedHeads","chain_subscribeNewHeads","chain_unsubscribeAllHeads","chain_unsubscribeFinalizedHeads","chain_unsubscribeNewHeads","childstate_getKeys","childstate_getStorage","childstate_getStorageHash","childstate_getStorageSize","grandpa_roundState","offchain_localStorageGet","offchain_localStorageSet","payment_queryInfo","rpc_methods","state_call","state_getKeys","state_getKeysPaged","state_getMetadata","state_getPairs","state_getReadProof","state_getRuntimeVersion","state_getStorage","state_getStorageHash","state_getStorageSize","state_queryStorage","state_queryStorageAt","state_subscribeRuntimeVersion","state_subscribeStorage","state_unsubscribeRuntimeVersion","state_unsubscribeStorage","system_accountNextIndex","system_addReservedPeer","system_chain","system_chainType","system_dryRun","system_health","system_localListenAddresses","system_localPeerId","system_name","system_networkState","system_nodeRoles","system_peers","system_properties","system_removeReservedPeer","system_version","chainHead_v1_body","chainHead_v1_call","chainHead_v1_follow","chainHead_v1_header","chainHead_v1_stopOperation","chainHead_v1_storage","chainHead_v1_continue","chainHead_v1_unfollow","chainHead_v1_unpin","chainSpec_v1_chainName","chainSpec_v1_genesisHash","chainSpec_v1_properties","sudo_unstable_p2pDiscover","sudo_unstable_version","transaction_v1_broadcast","transaction_v1_stop","transactionWatch_v1_submitAndWatch","transactionWatch_v1_unwatch","sudo_network_unstable_watch","sudo_network_unstable_unwatch","chainHead_unstable_finalizedDatabase"]}}
22:42:25 (Fiber #22) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","id":"1-2","method":"chainHead_v1_follow","params":[true]}
22:42:25 (Fiber #13) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","id":"1-2","result":"3tfjmBBFwSFvH8un27W9wqdhZ8K3fRn4ZK4UVmrHzuSU"}
22:42:29 (Fiber #34) INFO  runtime-polkadot Finalized block runtime ready. Spec version: 1002000. Size of `:code`: 1.5 MiB.
22:42:29 (Fiber #13) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","method":"chainHead_v1_followEvent","params":{"subscription":"3tfjmBBFwSFvH8un27W9wqdhZ8K3fRn4ZK4UVmrHzuSU","result":{"event":"initialized","finalizedBlockHashes":["0x90e0f0464643fa27d8943e5a1d06b18ee25daee1bb0d2d421954bb7c369e6dcd"],"finalizedBlockRuntime":{"type":"valid","spec":{"specName":"polkadot","implName":"parity-polkadot","specVersion":1002000,"implVersion":0,"transactionVersion":25,"apis":{"0xf78b278be53f454c":2,"0xed99c5acb25eedf5":3,"0xdf6acb689907609b":4,"0x40fe3ad401f8959a":6,"0xd2bc9897eed08f15":3,"0x18ef58a3b67ba770":1,"0x49eaaf1b548a0cb0":3,"0x91d5df18b0d2cf58":2,"0x2a5e924655399e60":1,"0x37c8bb1350a9a2a8":4,"0x37e397fc7c91f5e4":2,"0x17a6bc0d0062aeb3":1,"0xf3ff14d5ab527059":3,"0xbc9d89904f5b923f":1,"0xfbc577b9d747efd6":1,"0xcbca25e39f142387":2,"0xaf2c0297a23e6d3d":10,"0xab3c0572291feb8b":1,"0x687ad44ad37f03c2":1}}}}}}
22:42:29 (Fiber #35) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","id":"1-3","method":"chainHead_v1_header","params":["3tfjmBBFwSFvH8un27W9wqdhZ8K3fRn4ZK4UVmrHzuSU","0x90e0f0464643fa27d8943e5a1d06b18ee25daee1bb0d2d421954bb7c369e6dcd"]}
sub! {
  type: 'initialized',
  finalizedBlockHashes: [
    '0x90e0f0464643fa27d8943e5a1d06b18ee25daee1bb0d2d421954bb7c369e6dcd'
  ],
  finalizedBlockRuntime: {
    type: 'valid',
    spec: {
      specName: 'polkadot',
      implName: 'parity-polkadot',
      specVersion: 1002000,
      implVersion: 0,
      transactionVersion: 25,
      apis: [Object]
    }
  }
}
22:42:29 (Fiber #13) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","id":"1-3","result":"0xd3639eadb2a451492a5de1559b664ed854af823e2f12770839f5e671a77739c5de52ec04d9721bebfedcc9f6c81b919518c2107cecbf26a059aa578f07615fb2b505d15d0a4e28a6b59ec8130c9af6431ed4125c498f05a28fde249a284b8cb6b5f2e5ff0c0642414245b50103d70000009560091100000000543267a172aa16f7f2c1cded7ef6fd3b48989832f71225204e86fe354c56b74df5764f462f8ff075e802909f47021bc4736467bc53dc0730b684e6108084ea03cd9ea6bf6d8fe64fef57be8389439b2060ddd753e3b0a8780715797fe0976e0404424545468403837905c5291a4de90c7e8553695059fc679855e5c581f339404571fec0b22042054241424501013e39963542c432f263433c744c4de6d98425607919c6719279d21a43157f024de150c723e6729176ad074f0fc6190fc65366f75e0d659724a13b4de7c9720a8b"}
22:42:29 (Fiber #36) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","id":"1-4","method":"chainHead_v1_call","params":["3tfjmBBFwSFvH8un27W9wqdhZ8K3fRn4ZK4UVmrHzuSU","0x90e0f0464643fa27d8943e5a1d06b18ee25daee1bb0d2d421954bb7c369e6dcd","Metadata_metadata_versions",""]}
22:42:29 (Fiber #13) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","id":"1-4","result":{"result":"started","operationId":"3gTJQBu4wggd8kHGVCJ9vQJdFdf1Thuyvp2C5jNPcqyS"}}
22:42:29 (Fiber #13) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","method":"chainHead_v1_followEvent","params":{"subscription":"3tfjmBBFwSFvH8un27W9wqdhZ8K3fRn4ZK4UVmrHzuSU","result":{"event":"newBlock","blockHash":"0xbc02d7f8836dca19aa9f97692c883db8fec81729a2425099d89891f50fca31ab","parentBlockHash":"0x90e0f0464643fa27d8943e5a1d06b18ee25daee1bb0d2d421954bb7c369e6dcd","newRuntime":null}}}
sub! {
  type: 'newBlock',
  blockHash: '0xbc02d7f8836dca19aa9f97692c883db8fec81729a2425099d89891f50fca31ab',
  parentBlockHash: '0x90e0f0464643fa27d8943e5a1d06b18ee25daee1bb0d2d421954bb7c369e6dcd',
  newRuntime: null
}
22:42:29 (Fiber #13) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","method":"chainHead_v1_followEvent","params":{"subscription":"3tfjmBBFwSFvH8un27W9wqdhZ8K3fRn4ZK4UVmrHzuSU","result":{"event":"bestBlockChanged","bestBlockHash":"0xbc02d7f8836dca19aa9f97692c883db8fec81729a2425099d89891f50fca31ab"}}}
sub! {
  type: 'bestBlockChanged',
  bestBlockHash: '0xbc02d7f8836dca19aa9f97692c883db8fec81729a2425099d89891f50fca31ab'
}
22:42:29 (Fiber #13) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","method":"chainHead_v1_followEvent","params":{"subscription":"3tfjmBBFwSFvH8un27W9wqdhZ8K3fRn4ZK4UVmrHzuSU","result":{"event":"newBlock","blockHash":"0x145a9f1cb7b9f6bd0fc56f479a0eb4fa7fb985c1df2a9da7640ba441b0889eae","parentBlockHash":"0xbc02d7f8836dca19aa9f97692c883db8fec81729a2425099d89891f50fca31ab","newRuntime":null}}}
sub! {
  type: 'newBlock',
  blockHash: '0x145a9f1cb7b9f6bd0fc56f479a0eb4fa7fb985c1df2a9da7640ba441b0889eae',
  parentBlockHash: '0xbc02d7f8836dca19aa9f97692c883db8fec81729a2425099d89891f50fca31ab',
  newRuntime: null
}
22:42:29 (Fiber #13) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","method":"chainHead_v1_followEvent","params":{"subscription":"3tfjmBBFwSFvH8un27W9wqdhZ8K3fRn4ZK4UVmrHzuSU","result":{"event":"bestBlockChanged","bestBlockHash":"0x145a9f1cb7b9f6bd0fc56f479a0eb4fa7fb985c1df2a9da7640ba441b0889eae"}}}
sub! {
  type: 'bestBlockChanged',
  bestBlockHash: '0x145a9f1cb7b9f6bd0fc56f479a0eb4fa7fb985c1df2a9da7640ba441b0889eae'
}
22:42:29 (Fiber #13) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","method":"chainHead_v1_followEvent","params":{"subscription":"3tfjmBBFwSFvH8un27W9wqdhZ8K3fRn4ZK4UVmrHzuSU","result":{"event":"newBlock","blockHash":"0x33fc19c13eca50bf159b1cca88acbc32e43dcebef695df005958cd36df239100","parentBlockHash":"0x145a9f1cb7b9f6bd0fc56f479a0eb4fa7fb985c1df2a9da7640ba441b0889eae","newRuntime":null}}}
sub! {
  type: 'newBlock',
  blockHash: '0x33fc19c13eca50bf159b1cca88acbc32e43dcebef695df005958cd36df239100',
  parentBlockHash: '0x145a9f1cb7b9f6bd0fc56f479a0eb4fa7fb985c1df2a9da7640ba441b0889eae',
  newRuntime: null
}
22:42:29 (Fiber #13) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","method":"chainHead_v1_followEvent","params":{"subscription":"3tfjmBBFwSFvH8un27W9wqdhZ8K3fRn4ZK4UVmrHzuSU","result":{"event":"bestBlockChanged","bestBlockHash":"0x33fc19c13eca50bf159b1cca88acbc32e43dcebef695df005958cd36df239100"}}}
sub! {
  type: 'bestBlockChanged',
  bestBlockHash: '0x33fc19c13eca50bf159b1cca88acbc32e43dcebef695df005958cd36df239100'
}
22:42:29 (Fiber #13) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","method":"chainHead_v1_followEvent","params":{"subscription":"3tfjmBBFwSFvH8un27W9wqdhZ8K3fRn4ZK4UVmrHzuSU","result":{"event":"finalized","finalizedBlockHashes":["0xbc02d7f8836dca19aa9f97692c883db8fec81729a2425099d89891f50fca31ab"],"prunedBlockHashes":[]}}}
sub! {
  type: 'finalized',
  finalizedBlockHashes: [
    '0xbc02d7f8836dca19aa9f97692c883db8fec81729a2425099d89891f50fca31ab'
  ],
  prunedBlockHashes: []
}
22:42:29 (Fiber #13) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","method":"chainHead_v1_followEvent","params":{"subscription":"3tfjmBBFwSFvH8un27W9wqdhZ8K3fRn4ZK4UVmrHzuSU","result":{"event":"operationCallDone","operationId":"3gTJQBu4wggd8kHGVCJ9vQJdFdf1Thuyvp2C5jNPcqyS","output":"0x080e0000000f000000"}}}
22:42:29 (Fiber #38) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","id":"1-5","method":"chainHead_v1_call","params":["3tfjmBBFwSFvH8un27W9wqdhZ8K3fRn4ZK4UVmrHzuSU","0x90e0f0464643fa27d8943e5a1d06b18ee25daee1bb0d2d421954bb7c369e6dcd","Metadata_metadata_at_version","0x0f000000"]}
22:42:29 (Fiber #13) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","id":"1-5","result":{"result":"started","operationId":"4jVndw1241ZuAHej3o5QmYzLayDaJSA3CW7MabWRFPYw"}}
22:42:29 (Fiber #0) WARN  smoldot Smoldot is restarting...
22:42:29 (Fiber #39) INFO  smoldot Shutting down chain polkadot
22:42:29 (Fiber #13) ERROR json-rpc-provider -> polkadot {"_tag":"AlreadyDestroyedError"}
err! StopError: ChainHead stopped
    at Object.onAllFollowEventsNext (/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/@polkadot-api+substrate-client@0.1.1/node_modules/@polkadot-api/substrate-client/src/chainhead/chainhead.ts:83:21)
    at Object.next (/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/@polkadot-api+substrate-client@0.1.1/node_modules/@polkadot-api/substrate-client/src/internal-utils/subscriptions-manager.ts:18:30)
    at onMessage (/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/@polkadot-api+substrate-client@0.1.1/node_modules/@polkadot-api/substrate-client/src/client/createClient.ts:91:23)
    at onAbort (/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/@polkadot-api+json-rpc-provider-proxy@0.1.0/node_modules/@polkadot-api/json-rpc-provider-proxy/src/subscription-manager/chainHeadFollow.ts:50:7)
    at Array.forEach (<anonymous>)
    at Object.onAbort (/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/@polkadot-api+json-rpc-provider-proxy@0.1.0/node_modules/@polkadot-api/json-rpc-provider-proxy/src/subscription-manager/subscription-manager.ts:50:18)
    at <anonymous> (/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/@polkadot-api+json-rpc-provider-proxy@0.1.0/node_modules/@polkadot-api/json-rpc-provider-proxy/src/subscription-manager/subscription-manager.ts:80:38)
    at Array.forEach (<anonymous>)
    at Object.onAbort (/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/@polkadot-api+json-rpc-provider-proxy@0.1.0/node_modules/@polkadot-api/json-rpc-provider-proxy/src/subscription-manager/subscription-manager.ts:80:21)
    at onHalt (/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/@polkadot-api+json-rpc-provider-proxy@0.1.0/node_modules/@polkadot-api/json-rpc-provider-proxy/src/get-sync-provider.ts:71:27)
22:42:30 (Fiber #0) INFO  smoldot Smoldot restarted successfully.
22:42:30 (Fiber #42) INFO  smoldot Smoldot v2.0.25
22:42:31 (Fiber #45) INFO  smoldot Chain initialization complete for polkadot. Name: "Polkadot". Genesis hash: 0x91b1…90c3. Chain specification starting at: 0x9548…ebd3 (#20408370)
22:42:31 (Fiber #40) INFO  json-rpc-provider -> polkadot JSON RPC provider initialized.
22:42:31 (Fiber #50) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","id":"1-6","method":"chainHead_v1_stopOperation","params":["3tfjmBBFwSFvH8un27W9wqdhZ8K3fRn4ZK4UVmrHzuSU","4jVndw1241ZuAHej3o5QmYzLayDaJSA3CW7MabWRFPYw"]}
22:42:31 (Fiber #51) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","id":"1-7","method":"chainHead_v1_follow","params":[true]}
22:42:31 (Fiber #13) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","id":"1-6","result":null}
22:42:31 (Fiber #13) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","id":"1-7","result":"CJWZWGgdP9vaoPg6j2xxVHofW9e6P69WBaJF7gFAWnZH"}
22:42:31 (Fiber #53) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","id":"1-8","method":"chainHead_v1_unpin","params":["CJWZWGgdP9vaoPg6j2xxVHofW9e6P69WBaJF7gFAWnZH",["0x90e0f0464643fa27d8943e5a1d06b18ee25daee1bb0d2d421954bb7c369e6dcd"]]}
22:42:31 (Fiber #13) DEBUG json-rpc-provider -> polkadot {"jsonrpc":"2.0","id":"1-8","error":{"code":-32602,"message":"Invalid method parameter(s)."}}
/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/@polkadot-api+substrate-client@0.1.1/node_modules/@polkadot-api/substrate-client/src/client/createClient.ts:72
          ? cb.onError(new RpcError(error))
                       ^


RpcError: Invalid method parameter(s).
    at onMessage (/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/@polkadot-api+substrate-client@0.1.1/node_modules/@polkadot-api/substrate-client/src/client/createClient.ts:72:24)
    at onMessageProxy (/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/@polkadot-api+json-rpc-provider-proxy@0.1.0/node_modules/@polkadot-api/json-rpc-provider-proxy/src/get-sync-provider.ts:40:7)
    at _onMessageProxy (/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/@polkadot-api+json-rpc-provider-proxy@0.1.0/node_modules/@polkadot-api/json-rpc-provider-proxy/src/get-sync-provider.ts:95:11)
    at <anonymous> (/home/ryan/Documents/Repositories/substrate-connect/packages/light-client-experimental/dist/json-rpc-provider/index.mjs:109:40)
    at EffectPrimitive.effect_instruction_i0 (/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/effect@3.1.2/node_modules/effect/src/internal/core-effect.ts:102:14)
    at FiberRuntime.Sync (/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/effect@3.1.2/node_modules/effect/src/internal/fiberRuntime.ts:1076:22)
    at <anonymous> (/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/effect@3.1.2/node_modules/effect/src/internal/fiberRuntime.ts:1308:53)
    at f (/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/effect@3.1.2/node_modules/effect/src/internal/tracer.ts:93:19)
    at runLoop (/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/effect@3.1.2/node_modules/effect/src/internal/fiberRuntime.ts:1298:28)
    at evaluateEffect (/home/ryan/Documents/Repositories/substrate-connect/node_modules/.pnpm/effect@3.1.2/node_modules/effect/src/internal/fiberRuntime.ts:891:27) {
  code: -32602,
  data: undefined
}

Node.js v20.12.2
```

**Comment by ryanleecode** [comment]
![image](https://github.com/polkadot-api/polkadot-api/assets/21375952/d3c8bfb7-1cc6-4afa-a92b-94b988b6304b)

**Comment by voliva** [comment]
Thanks for reporting the issue.

Looking at the logs I'm not sure what's wrong though. Smoldot replies with `Invalid method parameters` for the request `1-8`, which is:

> {"jsonrpc":"2.0","id":"1-8","method":"chainHead_v1_unpin","params":["CJWZWGgdP9vaoPg6j2xxVHofW9e6P69WBaJF7gFAWnZH",["0x90e0f0464643fa27d8943e5a1d06b18ee25daee1bb0d2d421954bb7c369e6dcd"]]}

But those parameters look OK to me. Reading the JSON-RPC spec https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_v1_unpin.html, the first parameter is the token from the follow subscription, and the second is an array of hashes, and this one was the initial finalized block.

I'll investigate this a bit further, but it looks like it might be an issue on smoldot.

**Comment by voliva** [comment]
Fixed in #493
